app/controllers/checkout_controller.rb

```diff
@@ -142,7 +142,7 @@ def valid_payment_intent_provided?
   end
 
   def handle_redirect_from_stripe
-    if OrderWorkflow.new(@order).next && order_complete?
+    if OrderWorkflow.new(@order).next && order_complete? && @order.process_payments!
```

Comment:
So, we will move the order to complete and then process payments?
We need to move the order back fro complete if payment fails, right?

Reply:
I thought the order payment state was when we take payments. Is this changing? Do we keep the payment state, what for?

Reply:
Independently of the points above, I dont think this line can be here, I think this needs to be moved to a service. The relation between the order workflow and the taking of payment should be isolated in a service.

Reply:
We want to change when we take the payments, yes. Namely we want to be able to have payment processing be separate from the state of the order. We may be able to remove the payment state entirely in the future, but there's other callbacks (`before_transition to: :payment, do: :create_tax_charge!` and `after_transition to: :payment, do: :set_payment_amount!` that we'd have to consider as well, and I wanted to keep this PR as small as possible. 

> We need to move the order back fro complete if payment fails, right?

I don't think so, not unless we specifically want to cancel the order. If a first attempt at payment fails, we should keep the order in `complete` and allow the customer or a manager to add an additional payment.

Reply:
ok, that makes sense and sounds really good :+1: 

the only problem I see then is the fact that this logic should be in a service :+1:

Reply:
I also think (it's one of things that pop up to mind all the time) that the entire `#process_payments!` logic should be extracted and not hang from the omnipresent order model. But I agree with @andrewpbrett to be extra cautious and do that separately.

Reply:
> If a first attempt at payment fails, we should keep the order in complete and allow the customer or a manager to add an additional payment.

We don't want to do that for the checkout though, right? Currently if payment fails at the checkout the order is not completed, and we don't want to change that behavior?

Reply:
Yeah, that's a really good catch @Matt-Yorkley. I just pushed some changes to reflect that. It also means we no longer do this three part chained conditional so that might reduce the need to move things to a service in the immediate term... but there's still a lot of opportunity to refactor here. 

This highlights the difference between order processing in the shopfront and backoffice: in the shopfront flow, we check for stock availability at the beginning of each action, so we're confident that we can process the payment before moving the order to the complete state. In the backoffice, we complete the order first in order to reduce the stock, so that we don't take payment for an order that we don't have stock for. 

Now I'm wondering if instead of, or in addition to, this PR we could add the stock check to the backoffice flow instead of using the action of "completing the order" as a tool to accomplish the goal of checking stock...